src/main/java/org/ethereum/beacon/discovery/storage/KBucket.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.ethereum.beacon.discovery.storage;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.ethereum.beacon.discovery.schema.NodeRecord;
+import org.ethereum.beacon.discovery.storage.KBuckets.LivenessChecker;
+
+public class KBucket {
+
+  static final int K = 16;
+
+  private final LivenessChecker livenessChecker;
+  private final Clock clock;
+
+  private final List<BucketEntry> nodes = new ArrayList<>();
+
+  private Optional<BucketEntry> pendingNode = Optional.empty();
+
+  public KBucket(final LivenessChecker livenessChecker, final Clock clock) {
+    this.livenessChecker = livenessChecker;
+    this.clock = clock;
+  }
+
+  public List<NodeRecord> getAllNodes() {
+    return nodes.stream().map(BucketEntry::getNode).collect(Collectors.toList());
+  }
+
+  public List<NodeRecord> getLiveNodes() {
+    return nodes.stream()
+        .filter(node -> node.isLive())
+        .map(BucketEntry::getNode)
+        .collect(Collectors.toList());
+  }
+
+  public Optional<NodeRecord> getPendingNode() {
+    return pendingNode.map(BucketEntry::getNode);
+  }
+
+  public void offer(final NodeRecord node) {
+    performMaintenance();
+    getEntry(node)
+        .ifPresentOrElse(
+            existing -> updateExistingRecord(existing, node), () -> offerNewNode(node));
+  }
+
+  private void updateExistingRecord(final BucketEntry existing, final NodeRecord newRecord) {
+    if (existing.getNode().getSeq().compareTo(newRecord.getSeq()) >= 0) {
+      // New record isn't actually newer, do nothing.
+      return;
+    }
+    nodes.remove(existing);
+    final BucketEntry newEntry = new BucketEntry(livenessChecker, newRecord);
+    nodes.add(newEntry);
+    newEntry.checkLiveness(clock.millis());
+  }
+
+  private void offerNewNode(final NodeRecord node) {
+    if (isFull()) {
+      getLastNode().checkLiveness(clock.millis());
+      if (pendingNode.isEmpty()) {
+        livenessChecker.checkLiveness(node);
```

Comment:
would the intent of checking liveness here be to set the pendingNode if the incoming `node` is live? if that's the intent, it seems we might want to `pendingNode = new BucketEntry(livenessChecker, node);` here...

Reply:
The intent of the pending node is to hold one entry which is confirmed to be live and thus ready to be inserted immediately should a node be removed. So as we discover nodes that don't fit, we test their liveness but only actually store them as the pending node (and stop pinging new nodes) once we hear back from them (which comes via `onNodeContacted`).

Might rename `onNodeContacted` to `onLivenessConfirmed` actually.